CameraOverlay/overlay.py

```diff
@@ -1,20 +1,74 @@
-from picamera import PiCamera
 import time
-import paho.mqtt.client as mqtt
+from enum import Enum
 from abc import ABC, abstractmethod
+import cv2
+import numpy as np
+import paho.mqtt.client as mqtt
+
+try:
+	from picamera import PiCamera, PiRGBArray
+	ON_PI = True
+except (ImportError, RuntimeError):
+	ON_PI = False
+
+class Color(Enum):
+	# Remember, OpenCV uses BGR(A) not RGB(A)
+	white = (255, 255, 255, 255)
+	black = (0, 0, 0, 255)
+	blue = (255, 0, 0, 255)
+	green = (0, 255, 0, 255)
+	red = (0, 0, 255, 255)
+
+class Canvas():
+	""" A writeable image, for creating overlay content """
+
+	def __init__(self, width, height):
+		""" Initialises to plain black and transparent """
+		self.width = width
+		self.height = height
+		self.clear()
+
+	def clear(self):
+		self.img = np.zeros((self.height, self.width, 4), np.uint8)
+
+	def draw_text(self, text, coord, size=1.5, color=Color.black):
+		font = cv2.FONT_HERSHEY_SIMPLEX
+		thickness = round(size + 0.5)
+		cv2.putText(self.img, text, coord, font, size, color.value, thickness, cv2.LINE_AA)
 
+	def draw_rect(self, top_left, bottom_right, color=Color.black):
+		cv2.rectangle(self.img, top_left, bottom_right, color.value, thickness=cv2.FILLED)
+
+	def copy_to(self, dest):
+		""" Writes the contents of self.img over dest, accounting for transpaency
+		    Use this method to put the overlay contents over the video feed """
+		# Extract the alpha mask of the BGRA canvas, convert to BGR
+		blue, green, red, alpha = cv2.split(self.img)
+		img = cv2.merge((blue, green, red))
+		_, mask = cv2.threshold(alpha, 180, 255, cv2.THRESH_BINARY)
+
+		# Black-out the area behind the canvas on the destination
+		dest = cv2.bitwise_and(dest, dest, mask=cv2.bitwise_not(mask))
```

Comment:
wow. amazing. 💯

Reply:
omg this is super cool, applying bitwise operations irl ❤️